open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/utils/App2AppAuthUtils.java

```diff
@@ -0,0 +1,180 @@
+            RealmService realmService = IdentityExtensionsDataHolder.getInstance().getRealmService();
+            userRealm = realmService.getTenantUserRealm(tenantId);
+        }
+
+        return userRealm;
+    }
+
+    /**
+     * Retrieves the user ID associated with the provided username from the specified user realm.
+     *
+     * @param username the username for which to retrieve the user ID
+     * @param userRealm the user realm from which to retrieve the user ID
+     * @return the user ID associated with the username
+     * @throws UserStoreException if an error occurs while retrieving the user ID
+     */
+    public static String getUserIdFromUsername(String username, UserRealm userRealm) throws UserStoreException,
+            OpenBankingException  {
+
+        if (userRealm != null) {
+            AbstractUserStoreManager userStoreManager = (AbstractUserStoreManager) userRealm.getUserStoreManager();
+            return userStoreManager.getUserIDFromUserName(username);
+        } else {
+            throw new OpenBankingException("UserRealm service can not be null.");
+        }
+    }
+
+    /**
+     * Retrieves the registered device associated with the specified device ID and user ID.
+     * TODO: Optimise this code to retrieve device by did and validate userID.
+     * Github issue :{<a href="https://github.com/wso2-extensions/identity-outbound-auth-push/issues/144">...</a>}
+     *
+     * @param deviceId the ID of the device to retrieve
+     * @param userId the ID of the user who owns the device
+     * @return the registered device associated with the specified IDs
+     * @throws PushDeviceHandlerServerException if an error occurs on the server side while handling the device
+     * @throws IllegalArgumentException if the provided device identifier does not exist
+     * @throws PushDeviceHandlerClientException if an error occurs on the client side while handling the device
+     */
+    public static Device getRegisteredDevice(String deviceId, String userId, DeviceHandler deviceHandler)
+            throws PushDeviceHandlerServerException, IllegalArgumentException,
+            PushDeviceHandlerClientException, OpenBankingException {
+
+        /*
+         It is important to verify the device is registered under the given user
+         as public key is associated with device not the user.
+         */
+        List<Device> deviceList = deviceHandler.listDevices(userId);
+        //If a device registered under the given user matches the specified deviceId return the device
+        Optional<Device> optionalDevice = deviceList.stream()
+                .filter(device -> StringUtils.equals(device.getDeviceId(), deviceId))
+                .findFirst();
+        //If no device found throw exception
+        Device device =  optionalDevice.orElseThrow(() ->
+                new OpenBankingException("Provided Device ID doesn't match any device registered under user."));
```

Comment:
```suggestion
        //If a device registered under the given user matches the specified deviceId return the device
        Device device = deviceHandler.listDevices(userId).stream()
                .filter(device -> StringUtils.equals(device.getDeviceId(), deviceId))
                .findFirst()
                //If no device found throw exception
                .orElseThrow(() ->
                    new OpenBankingException("Provided Device ID doesn't match any device registered under user."));
```

Reply:
Resolved in [260e508](https://github.com/wso2/financial-open-banking/pull/67/commits/260e508da50ed60b02dc3fd2f2d27107b59691d2).